internal/tools/aws/filestore_multitenant.go

```diff
@@ -0,0 +1,238 @@
+	}
+}
+
+// Provision completes all the steps necessary to provision an S3 multitenant
+// filestore.
+func (f *S3MultitenantFilestore) Provision(store model.InstallationDatabaseStoreInterface, logger log.FieldLogger) error {
+	err := f.s3FilestoreProvision(store, logger)
+	if err != nil {
+		return errors.Wrap(err, "failed to provision AWS multitenant S3 filestore")
+	}
+
+	return nil
+}
+
+// Teardown removes all AWS resources related to a shared S3 filestore.
+func (f *S3MultitenantFilestore) Teardown(keepData bool, store model.InstallationDatabaseStoreInterface, logger log.FieldLogger) error {
+	awsID := CloudID(f.installationID)
+
+	logger = logger.WithFields(log.Fields{
+		"awsID":          awsID,
+		"filestore-type": "s3-multitenant",
+	})
+	logger.Info("Tearing down AWS S3 filestore")
+
+	bucketName, err := f.getMultitenantBucketName(store)
+	if err != nil {
+		return errors.Wrap(err, "failed to find multitenant bucket")
+	}
+
+	logger = logger.WithField("s3-bucket-name", bucketName)
+
+	err = f.awsClient.iamEnsureUserDeleted(awsID, logger)
+	if err != nil {
+		return errors.Wrap(err, "failed to delete AWS IAM user")
+	}
+
+	err = f.awsClient.secretsManagerEnsureIAMAccessKeySecretDeleted(awsID, logger)
+	if err != nil {
+		return errors.Wrap(err, "failed to delete IAM access key secret")
+	}
+
+	if keepData {
+		logger.Info("AWS S3 bucket was left intact due to the keep-data setting of this server")
+		return nil
+	}
+
+	err = f.awsClient.S3EnsureBucketDirectoryDeleted(bucketName, f.installationID, logger)
+	if err != nil {
+		return errors.Wrap(err, "unable to ensure that AWS S3 filestore was deleted")
+	}
+
+	logger.Debug("AWS multitenant S3 filestore was deleted")
+	return nil
+}
+
+// GenerateFilestoreSpecAndSecret creates the k8s filestore spec and secret for
+// accessing the shared S3 bucket.
+func (f *S3MultitenantFilestore) GenerateFilestoreSpecAndSecret(store model.InstallationDatabaseStoreInterface, logger log.FieldLogger) (*mmv1alpha1.Minio, *corev1.Secret, error) {
+	awsID := CloudID(f.installationID)
+
+	logger = logger.WithFields(log.Fields{
+		"awsID":          awsID,
+		"filestore-type": "s3-multitenant",
+	})
+	logger.Debug("Generating S3 multitenant filestore information")
+
+	bucketName, err := f.getMultitenantBucketName(store)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to find multitenant bucket")
+	}
+
+	logger = logger.WithField("s3-bucket-name", bucketName)
+
+	iamAccessKey, err := f.awsClient.secretsManagerGetIAMAccessKey(awsID, logger)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	filestoreSecretName := fmt.Sprintf("%s-iam-access-key", f.installationID)
+	filestoreSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: filestoreSecretName,
+		},
+		StringData: map[string]string{
+			"accesskey": iamAccessKey.ID,
+			"secretkey": iamAccessKey.Secret,
+		},
+	}
+
+	S3RegionURL := S3URL
+	awsRegion := *f.awsClient.config.Region
+	if awsRegion != "" && awsRegion != "us-east-1" {
```

Comment:
Why the exception for `us-east-1`? Is `s3.us-east-1.amazonaws.com` invalid or something?